src/api.cc

```diff
@@ -963,3 +1015,138 @@ InstallResult AkliteClient::SetSecondaries(const std::vector<SecondaryEcu>& ecus
   secondary_hwids_ = std::move(hwids);
   return InstallResult{InstallResult::Status::Ok, ""};
 }
+
+static Json::Value checkAndGetRootMeta(const std::shared_ptr<aklite::tuf::Repo>& device_tuf_repo,
+                                       const boost::filesystem::path& bundle_tuf_repo_path) {
+  auto latest_root{device_tuf_repo->GetRoot(-1)};
+  Uptane::ImageRepository repo;
+  const int max_version{std::numeric_limits<int>::max()};
+  int version{0};
+  if (!latest_root.empty()) {
+    repo.initRoot(Uptane::RepositoryType::Image(), latest_root);
+    version = repo.getRoleVersion(Uptane::Role::Root());
+  }
+
+  aklite::tuf::LocalRepoSource bundle_tuf_repo{"offline-bundle-tuf-repo", bundle_tuf_repo_path.string()};
+  do {
+    try {
+      ++version;
+      latest_root = bundle_tuf_repo.FetchRoot(version);
+      if (repo.rootVersion() > 0) {
+        repo.verifyRoot(latest_root);
+      } else {
+        repo.initRoot(Uptane::RepositoryType::Image(), latest_root);
+      }
+    } catch (const aklite::tuf::MetadataNotFoundException& exc) {
+      break;
+    }
+  } while (version < max_version);
+
+  return Utils::parseJSON(latest_root);
+}
+
+static Json::Value checkAndGetBundleMeta(const std::shared_ptr<aklite::tuf::Repo>& device_tuf_repo,
+                                         const boost::filesystem::path& bundle_tuf_repo_path) {
+  const auto root_meta{checkAndGetRootMeta(device_tuf_repo, bundle_tuf_repo_path)};
+
+  std::set<std::string> targets_key_ids;
+  for (const auto& key : root_meta["signed"]["roles"]["targets"]["keyids"]) {
+    targets_key_ids.insert(key.asString());
+  }
+
+  std::map<std::string, PublicKey> targets_pub_keys;
+  for (Json::ValueConstIterator it = root_meta["signed"]["keys"].begin(); it != root_meta["signed"]["keys"].end();
+       ++it) {
+    const std::string keyid{it.key().asString()};
+    if (targets_key_ids.count(keyid) == 1) {
+      targets_pub_keys.insert({keyid, PublicKey{*it}});
+    }
+  }
+
+  const int targets_sign_threshold{root_meta["signed"]["roles"]["targets"]["threshold"].asInt()};
+
+  int found_valid_sig_number{0};
+  const auto bundle_meta_path{bundle_tuf_repo_path / "bundle-targets.json"};
+  if (!boost::filesystem::exists(bundle_meta_path)) {
+    throw BundleMetaError(BundleMetaError::Type::NotFound,
+                          "Failed to find the bundle metadata: " + bundle_meta_path.string());
```

Comment:
Looking at this change I am under the impression that it _requires_ a `bundle-targets.json` to be present in a bundle.

But, there are already bundles used generated and used by our users without this file.
Don't you think we need to continue supporting those bundles?

Reply:
We do support them by the previous LmP/aklite versions until v94. If a device is already running the code with this check it means that a device is on v94. So, if the bundle was generated earlier it means that it has outdated TUF meta and target(s), and the update will fail anyway.

Reply:
Hmm... interesting logic.
I think I agree to it; and my hope is that there is no leak in it.

Reply:
I don't have really strong opinion about it, just don't see real-life use-case for it. Also, it is kind of mechanism to enforce customers to update fioctl to the version that supports uptane-line bundles.